File: shared/models/target_asset.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

MAX_ASSET_SEARCH = 200

ASSET_STATES: tuple[str, ...] = ("all", "current", "new", "gone")
ASSET_SORTS: tuple[str, ...] = ("name", "first_seen", "last_seen", "scans", "status")


class TargetAssetRow(BaseModel):
    name: str
    is_active: bool = False
    is_wildcard: bool = False
    resolved_ips: list[str] = Field(default_factory=list)
    cname: str | None = None
    sources: list[str] = Field(default_factory=list)
    scan_count: int = 0
    first_seen: datetime
    last_seen: datetime
    last_scan_id: uuid.UUID
    current: bool = True
    is_new: bool = False
    status_code: int | None = None
    title: str | None = None
    webserver: str | None = None
    tech: list[str] = Field(default_factory=list)
    ip: str | None = None
    asn_org: str | None = None
    is_cdn: bool = False
    cdn_name: str | None = None
    screenshot_path: str | None = None


class TargetAssetFilter(BaseModel):
    search: str | None = Field(default=None, max_length=MAX_ASSET_SEARCH)
    state: str = Field(default="all", max_length=10)
    live: bool = False
    sort: str = Field(default="name", max_length=20)
    order: str = Field(default="asc", max_length=4)
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0, le=100_000_000)


class TargetAssetFacets(BaseModel):
    total: int = 0
    current: int = 0
    new: int = 0
    gone: int = 0
    live: int = 0
    baseline: bool = False
    latest_scan_id: uuid.UUID | None = None


class TargetAssetPage(BaseModel):
    items: list[TargetAssetRow] = Field(default_factory=list)
    total: int = 0
    facets: TargetAssetFacets = Field(default_factory=TargetAssetFacets)

File: frontend/src/lib/types/target-asset.ts
export const ASSET_STATES = ['all', 'current', 'new', 'gone'] as const;
export type AssetState = (typeof ASSET_STATES)[number];

export const ASSET_SORTS = [
	{ key: 'name', label: 'Host' },
	{ key: 'status', label: 'HTTP status' },
	{ key: 'first_seen', label: 'First seen' },
	{ key: 'last_seen', label: 'Last seen' },
	{ key: 'scans', label: 'Scans seen in' }
] as const;
export type AssetSort = (typeof ASSET_SORTS)[number]['key'];

export interface TargetAssetRow {
	name: string;
	is_active: boolean;
	is_wildcard: boolean;
	resolved_ips: string[];
	cname: string | null;
	sources: string[];
	scan_count: number;
	first_seen: string;
	last_seen: string;
	last_scan_id: string;
	current: boolean;
	is_new: boolean;
	status_code: number | null;
	title: string | null;
	webserver: string | null;
	tech: string[];
	ip: string | null;
	asn_org: string | null;
	is_cdn: boolean;
	cdn_name: string | null;
	screenshot_path: string | null;
}

export interface TargetAssetFilter {
	search?: string | null;
	state?: AssetState;
	live?: boolean;
	sort?: AssetSort;
	order?: 'asc' | 'desc';
	limit?: number;
	offset?: number;
}

export interface TargetAssetFacets {
	total: number;
	current: number;
	new: number;
	gone: number;
	live: number;
	baseline: boolean;
	latest_scan_id: string | null;
}

export interface TargetAssetPage {
	items: TargetAssetRow[];
	total: number;
	facets: TargetAssetFacets;
}
